Predicate filter proposal

ark-konopacki opened the pull request.

jmoody commented on calabash-cucumber/lib/calabash-cucumber/map.rb:
```diff
@@ -82,6 +82,22 @@ def self.raw_map(query, method_name, *method_args)
+          end
+
+        end
+
```
Let's extract this to a method and add some rspec tests.

ark-konopacki commented:
@jmoody I moved implementation to new method and added two tests for that new method. I'm not super happy with second one since it can be `false positive`. Can you check?

jmoody commented on calabash-cucumber/spec/lib/map_spec.rb:
```diff
@@ -40,5 +40,36 @@
     end
   end
+
+  describe ".raw_map.correct_predicate" do
```
+100 for rspec tests.

jmoody commented on calabash-cucumber/lib/calabash-cucumber/map.rb:
```diff
@@ -143,6 +144,40 @@ def self.assert_map_results(map_results, msg)
+            end
+          end
+          if !correct
+            fail "Incorrect predicate used, valid selectors are: #{predicate}"
```
Don't use `fail` here.  Raise a RuntimeError.  We don't need to take a screenshot.

jmoody commented on calabash-cucumber/lib/calabash-cucumber/map.rb:
```diff
@@ -143,6 +144,40 @@ def self.assert_map_results(map_results, msg)
+          if !query.match(/:'.*'/).nil?
+            true
+          else
+            fail 'Incorrect query format please check query string'
```
Ditto.

jmoody commented on calabash-cucumber/lib/calabash-cucumber/map.rb:
```diff
@@ -143,6 +144,40 @@ def self.assert_map_results(map_results, msg)
+      # Evaluating whether a query contain correct predicate selector
+      # returns true if query include correct predicate selector
+      def self.correct_predicate?(query)
+        if !query.match(/{.*}/).nil?
```
Prefer this pattern:

```
def self.query_has_predicate?(query)
  query.match(/{.*}/)
end

# Expect implies an exception will be raised.
def self.expect_valid_predicate(query)
   match = self.query_has_predicate?(query)
   return true if match.nil?

   ...
end
```

jmoody replied:
There are actually two conditions for a valid predicate:

1. The predicate keyword is supported (CONTAINS, LIKE, etc).
2. The right-hand side of the predicate is valid (not missing a `'`).

jmoody commented on calabash-cucumber/lib/calabash-cucumber/map.rb:
```diff
@@ -143,6 +144,40 @@ def self.assert_map_results(map_results, msg)
+        if !query.match(/{.*}/).nil?
+          str = query.match(/{.*}/)[0]
+          correct = false
+          predicate = %w(BEGINSWITH CONTAINS ENDSWITH LIKE MATCHES)
```
Do not use `%w`; use `[ ]`.

jmoody replied:
Maybe extract these to a ruby constant so you can share them with the rspec tests.

jmoody commented on calabash-cucumber/lib/calabash-cucumber/map.rb:
```diff
@@ -143,6 +144,40 @@ def self.assert_map_results(map_results, msg)
+      # returns true if query include : character and two ' characters
+      # returns true if query does not include : character, like query("label")
+      # TODO: We can add additional verification in future
+      def self.correct_format?(query)
```
Rename to something like:

```
def self.right_hand_side_of_predicate_valid?(query)
```

jmoody commented:
There are some more test cases:

```
# Case insensitive matching.
LIKE[c]

# Diacritic insensitive matching
LIKE[d]

# Case and Diacritic insensitive matching
LIKE[cd]
```

We should have a test for those examples to ensure our `/{.*}/` regex works correctly.

I think there there is also the `as` case that we need to handle.

There is an example of this floating around somewhere.

ark-konopacki commented:
@jmoody there is two methods now:
`correct_predicate?` check selector and if right part have two `'` chars
and 
`correct_format?` to check two `'` chars in regular queries like `"label marked:'label'"`
Still more tests needed.

jmoody commented:
Can you rebase against the latest develop?
